update sepolia lc + runbook (#18)
cosmos/eureka-ops: runbooks/recover-expired-light-client.md

# Runbook — recover an expired SP1 light client

This runbook explains how to recover an **expired `SP1ICS07Tendermint` light client** (an IBC light client on Ethereum that tracks a Cosmos chain). It is written for someone with no prior context: it covers what the problem is, the systems involved, the exact commands, and every gotcha we hit doing it for real.

> For the *normal* (non-expired) light-client migration ceremony, see [`upgrade-light-client.md`](./upgrade-light-client.md). This runbook is the **expired** variant — it adds a step to regenerate fresh trusted state, then reuses the same deploy + migrate flow.

Reviewer: (Side note) So we use the same trusted state when the client is not expired. Interesting, maybe we should consider always using a new trusted state.

---

## 1. The problem

An `SP1ICS07Tendermint` contract verifies a Cosmos chain's Tendermint consensus using SP1 zero-knowledge proofs. Like any Tendermint light client it has a **trusting period** (typically 2/3 of the Cosmos chain's unbonding period). If no `updateClient` lands within that window, the latest trusted consensus state ages out and the client becomes **expired** — it can no longer be updated by a normal proof, because there is no recent-enough trusted state to verify against.

You **cannot** revive an expired client with a normal `updateClient`. The fix is to **deploy a freshly-initialised client** with a recent trusted header and **repoint the existing client ID** at it:

1. **Generate fresh trusted state** (new `trustedClientState` + `trustedConsensusStateHash`) from a running proof-api and write it into the deployment JSON.
2. **Deploy a new `SP1ICS07Tendermint` contract** initialised with that fresh state (existing vkeys/verifier reused).
3. **Migrate** the client ID to point at the new contract (`ics26Router.migrateClient`), via the timelock + Gnosis Safe.

> > Why step 1 is special for an *expired* client: the pre-existing `just deploy-update-light-client-state` reads trusted state from the **on-chain contract** (`getClientState()`), which is stale/expired and therefore useless. The `deploy-fresh-light-client-state` recipe instead pulls **fresh** state from the proof-api.

Reviewer: ditto: use github alert notation

```suggestion
> [!NOTE]
> Why step 1 is special for an *expired* client: the pre-existing `just deploy-update-light-client-state` reads trusted state from the **on-chain contract** (`getClientState()`), which is stale/expired and therefore useless. The `deploy-fresh-light-client-state` recipe instead pulls **fresh** state from the proof-api.
```

---

## 2. Systems involved

- **`eureka-ops`** (this repo) — source of truth for deployments; operations run via `just`. The target lives in `deployments/<env>/<chainId>.json` under `light_clients["<idx>"]`.
- **`solidity-ibc-eureka`** — upstream Solidity contracts + the `operator` binary. The deployed contracts were compiled from a tagged version of this repo (pinned in `package.json` as `@cosmos/solidity-ibc-eureka`).
- **The proof-api (a.k.a. eureka-relayer)** — a gRPC service that builds unsigned IBC txs. Its `CreateClient` method queries the Cosmos RPC and returns the **creation calldata** for a fresh `SP1ICS07Tendermint` (no proof generated). We use it as the source of fresh trusted state. It runs in k8s (see step 1 for access).

### ⚠️ Version drift — verify against the live chain, not local source

> The version the contracts were **deployed** from can differ from what the repo currently **compiles** and from the **running** proof-api image. Concretely, on the testnet recovery the live `ICS26Router` used OZ `AccessControl` while the pinned `solidity-v2.0.0` source uses `AccessManaged`. **Always confirm access control / addresses against the live contract** (`cast call`), not the checked-out source.

Reviewer: This is not true. Testnet version should be v2.0.0. v3.0.0 uses access manager.

```suggestion
The version the contracts were **deployed** from can differ from what the repo currently **compiles** and from the **running** proof-api image. Concretely, on the testnet recovery the live `ICS26Router` used OZ `AccessControl` while the pinned `solidity-v3.0.0` source uses `AccessManaged`. **Always confirm access control / addresses against the live contract** (`cast call`), not the checked-out source.
```

> The version the contracts were **deployed** from can differ from what the repo currently **compiles** and from the **running** proof-api image. Concretely, on the testnet recovery the live `ICS26Router` used OZ `AccessControl` while the pinned `solidity-v2.0.0` source uses `AccessManaged`. **Always confirm access control / addresses against the live contract** (`cast call`), not the checked-out source.

Reviewer: nit: might be better to use github alerts syntax here, since it looks better in the reader:

```suggestion
> [!WARNING]  
> The version the contracts were **deployed** from can differ from what the repo currently **compiles** and from the **running** proof-api image. Concretely, on the testnet recovery the live `ICS26Router` used OZ `AccessControl` while the pinned `solidity-v2.0.0` source uses `AccessManaged`. **Always confirm access control / addresses against the live contract** (`cast call`), not the checked-out source.
```

---

## 3. Roles

| Role         | Who |
|--------------|-----|
| Facilitator  | runs the recipes, deploys the new contract, gathers Safe-tx params |
| Signer(s)    | Gnosis Safe owner(s) who sign the schedule + execute Safe txs |
| Notekeeper   | records addresses / hashes for the PR |

---

## 4. Prerequisites

- Tooling: `foundry` (forge/cast/chisel), `bun`, `just`, `jq`, `fzf`, **`grpcurl`**, **`python3`**, `kubectl`.

> - **Proof-api reachable** for step 1 (k8s port-forward — see step 1).

Reviewer: I recommend linking the markdown for step 1 here. I was confused since sections are numbered in this doc, but step numbers are introduced later. So I actually jumped to Section 1 (instead of Step 1)

> - A **deploy signer with gas** for step 2 (any funded EOA; need not be a Safe owner — deploying is permissionless).

Reviewer: ditto: link to step 2

> - A **Safe owner** + the Safe for step 3.

Reviewer: ditto: link to step 3

> **Signing note:** the `just` deploy/timelock recipes sign via `forge`, which supports **Ledger, raw private key, mnemonic, or an encrypted keystore** — but **not MetaMask** (no CLI bridge). For step 3 the Safe transactions are signed in the **Safe UI** (MetaMask works there) or verified via the repo's EIP-712 hash recipes.

---

## 5. Step 0 — point the repo at the right deployment (`.eureka-env`)

`just` auto-loads `.eureka-env` (gitignored). Minimal example for Sepolia:

```
EUREKA_ENVIRONMENT=testnet
EUREKA_CHAIN=11155111
ETH_RPC=<a Sepolia RPC URL>

# Deploy signer (step 2). Pick ONE:
#  - Ledger:      set SENDER=0x... and leave PRIVATE_KEY unset
#  - Private key: set PRIVATE_KEY=0x... (any value here OVERRIDES the Ledger path)
SENDER=0x...
# PRIVATE_KEY=0x...
# ETHERSCAN_API_KEY=...   # only needed for --verify to succeed; deploy + JSON write happen regardless

# Step 1 proof-api module identifiers (see step 1 for how to find them):
SRC_CHAIN=<cosmos-chain-id>
DST_CHAIN=<eth-chain-id>       # decimal chain id, e.g. 11155111 for Sepolia
PROOF_TYPE=groth16            # MUST match the existing client's zk algorithm

# Step 3 Safe (the Safe that holds PROPOSER on the timelock):

> SAFE_ADDRESS=0x...

Reviewer: Maybe these should be added to `.eureka-env.example`, and you could perhaps just link to that?

```

> > ⚠️ **`PRIVATE_KEY` footgun:** setting it to *any* value (even zeros) makes the recipes use `--private-key` instead of the Ledger. Leave it commented to use a Ledger.

Reviewer: nit: use github alerts syntax

```suggestion
> [!WARNING]  
> ⚠️ **`PRIVATE_KEY` footgun:** setting it to *any* value (even zeros) makes the recipes use `--private-key` instead of the Ledger. Leave it commented to use a Ledger.
```

`just info-env` prints the resolved settings (private key redacted) — sanity-check the chain, RPC, and that the broadcast flags show the signer you expect.

---

## 6. Step 1 — generate fresh trusted state

### 6a. Reach the proof-api
The proof-api runs in k8s. Port-forward it to `localhost:3000`:

```
kubectl --context <eks-cluster-context> \
  -n <namespace> port-forward pod/<proof-api-pod> 3000:3000
```

> > 🔒 The concrete cluster context (ARN), namespace, and pod naming convention are internal infra and live in the internal docs, not this public repo.

Reviewer: ditto: use github alert syntax

```suggestion
> [!NOTE]
> 🔒 The concrete cluster context (ARN), namespace, and pod naming convention are internal infra and live in the internal docs, not this public repo.
```

Find the pod with `kubectl --context <eks-cluster-context> -n <namespace> get pods | grep relayer-api`. Confirm it's up: `grpcurl -plaintext localhost:3000 list` should list `relayer.RelayerService` (v0.7.x) or `proofapi.ProofApiService` (v0.8.x).

### 6b. Find `SRC_CHAIN` / `DST_CHAIN`
These must **exactly match** the proof-api's configured module identifiers — **not** necessarily the on-chain chain-id. Source of truth is the relayer config (internal repo + config path — see internal docs): find the `cosmos_to_eth` module whose `ics26_address` matches your `ics26Router.proxy`. `src_chain` is the Cosmos id (e.g. `provider`), `dst_chain` is the **decimal** Eth chain id (e.g. `11155111`).

You can also probe a candidate pair (metadata only, no proof):
```
grpcurl -plaintext -d '{"src_chain":"<src>","dst_chain":"<dst>"}' localhost:3000 relayer.RelayerService/Info
```
A configured pair does *not* return `Module not found`.

### 6c. Run the recipe
```
just deploy-fresh-light-client-state      # enter the client ID when prompted, confirm y
```

> > ⏱️ **`CreateClient` is slow (~3–4 min)** — the relayer fetches headers/validator sets from the Cosmos RPC. The recipe sets **no grpcurl timeout**, so just wait. **Do not abort and retry**: a `CreateClient` keeps running server-side for the full duration even after you disconnect, and blocks subsequent calls. The k8s port-forward survives the long request.

Reviewer: nit: use github alert syntax:

```suggestion
> [!NOTE] 
> ⏱️ **`CreateClient` is slow (~3–4 min)** — the relayer fetches headers/validator sets from the Cosmos RPC. The recipe sets **no grpcurl timeout**, so just wait. **Do not abort and retry**: a `CreateClient` keeps running server-side for the full duration even after you disconnect, and blocks subsequent calls. The k8s port-forward survives the long request.
```

> > ⏱️ **`CreateClient` is slow (~3–4 min)** — the relayer fetches headers/validator sets from the Cosmos RPC. The recipe sets **no grpcurl timeout**, so just wait. **Do not abort and retry**: a `CreateClient` keeps running server-side for the full duration even after you disconnect, and blocks subsequent calls. The k8s port-forward survives the long request.

Reviewer: didn't know it took 3-4 minutes

Author: yeah, I didn't dig into why this is happening. it might be dev environment specific 🤷

### 6d. Verify
```
git diff deployments/<env>/<chainId>.json
```
Only `light_clients["<idx>"].trustedClientState` and `.trustedConsensusStateHash` should change. Confirm the client state's latest height is recent.

---

## 7. Step 2 — deploy the new contract

```
just deploy-light-client      # enter the client ID, confirm "deploy a copy" with y
```
This deploys a new `SP1ICS07Tendermint` with the fresh state from step 1, **reusing** the existing vkeys + verifier, and writes the new `implementation` (and `verifier`) into the deployment JSON.

### ⚠️ Gotcha: OpenZeppelin version pin

> The `solidity-v2.0.0` contracts import `ReentrancyGuardTransientUpgradeable.sol`, which OpenZeppelin **removed from `@openzeppelin/contracts-upgradeable` after 5.4.0**. Both `@openzeppelin/contracts` and `@openzeppelin/contracts-upgradeable` are therefore pinned to exactly **`5.4.0`** in `package.json` (the newest version that still ships that file). Don't loosen them to a range — `^5.3.x` floats to 5.5+ and the build fails with *"ReentrancyGuardTransientUpgradeable.sol not found"*. If you hit that error, your `node_modules` floated past the pin; re-run `bun install`.

Reviewer: I think the better warning here would be always to use the same openzeppelin version that is used by the solidity contracts?

Author: v2.0.0's declares a range (^5.3.0), which today resolves to 5.6.x, where ReentrancyGuardTransientUpgradeable.sol no longer exists. So "use the same OZ version as the contracts" is what breaks the build. The 5.4.0 pin is the workaround until we move to v3.x.

> The `solidity-v2.0.0` contracts import `ReentrancyGuardTransientUpgradeable.sol`, which OpenZeppelin **removed from `@openzeppelin/contracts-upgradeable` after 5.4.0**. Both `@openzeppelin/contracts` and `@openzeppelin/contracts-upgradeable` are therefore pinned to exactly **`5.4.0`** in `package.json` (the newest version that still ships that file). Don't loosen them to a range — `^5.3.x` floats to 5.5+ and the build fails with *"ReentrancyGuardTransientUpgradeable.sol not found"*. If you hit that error, your `node_modules` floated past the pin; re-run `bun install`.

Reviewer: ditto: wrong version

```suggestion
### ⚠️ Gotcha: OpenZeppelin version pin
The `solidity-v3.0.0` contracts import `ReentrancyGuardTransientUpgradeable.sol`, which OpenZeppelin **removed from `@openzeppelin/contracts-upgradeable` after 5.4.0**. Both `@openzeppelin/contracts` and `@openzeppelin/contracts-upgradeable` are therefore pinned to exactly **`5.4.0`** in `package.json` (the newest version that still ships that file). Don't loosen them to a range — `^5.3.x` floats to 5.5+ and the build fails with *"ReentrancyGuardTransientUpgradeable.sol not found"*. If you hit that error, your `node_modules` floated past the pin; re-run `bun install`.
```

Author: This one is correct as-is, actually: the v2.0.0 contracts do import ReentrancyGuardTransientUpgradeable.sol

ICS26Router.sol#L13-L14 at solidity-v2.0.0. It's v3.0.0 that switched to the non-upgradeable ReentrancyGuardTransient

> The `solidity-v2.0.0` contracts import `ReentrancyGuardTransientUpgradeable.sol`, which OpenZeppelin **removed from `@openzeppelin/contracts-upgradeable` after 5.4.0**. Both `@openzeppelin/contracts` and `@openzeppelin/contracts-upgradeable` are therefore pinned to exactly **`5.4.0`** in `package.json` (the newest version that still ships that file). Don't loosen them to a range — `^5.3.x` floats to 5.5+ and the build fails with *"ReentrancyGuardTransientUpgradeable.sol not found"*. If you hit that error, your `node_modules` floated past the pin; re-run `bun install`.

Reviewer: ditto: use github alert syntax

```suggestion
> [!IMPORTANT]  
> The `solidity-v2.0.0` contracts import `ReentrancyGuardTransientUpgradeable.sol`, which OpenZeppelin **removed from `@openzeppelin/contracts-upgradeable` after 5.4.0**. Both `@openzeppelin/contracts` and `@openzeppelin/contracts-upgradeable` are therefore pinned to exactly **`5.4.0`** in `package.json` (the newest version that still ships that file). Don't loosen them to a range — `^5.3.x` floats to 5.5+ and the build fails with *"ReentrancyGuardTransientUpgradeable.sol not found"*. If you hit that error, your `node_modules` floated past the pin; re-run `bun install`.
```

```
git diff deployments/<env>/<chainId>.json    # light_clients["<idx>"].implementation -> new address
cast call <newImpl> 'getClientState()(bytes)' --rpc-url <eth-rpc>   # optional sanity check
```

---

## 8. Step 3 — migrate the client ID (timelock + Gnosis Safe)

### 8a. Confirm who holds the migrator role
`migrateClient` on the router is access-controlled. Confirm the controller chain on-chain (don't trust local source — see §2). For the testnet deployment it was: a **`TimelockController`** (`ics26Router.timelockAdmin`) fronted by a **Gnosis Safe** that holds `PROPOSER`/`EXECUTOR`/`CANCELLER`. Useful checks:
```
cast call <timelockAdmin> 'getMinDelay()(uint256)' --rpc-url <eth-rpc>            # it's a TimelockController
cast call <safe> 'getThreshold()(uint256)' --rpc-url <eth-rpc>                    # it's a Safe
# is the timelock allowed to migrate? simulate from it (should NOT revert):
cast call <router> 'migrateClient(string,(string,bytes[]),address)' <clientId> "(<cpClientId>,[<merklePrefix...>])" <newImpl> --from <timelockAdmin> --rpc-url <eth-rpc>
```
If instead an EOA holds the role, use the direct `just ops-migrate-light-client` and skip the Safe ceremony.

> ### 8b. `SAFE_ADDRESS` must point at the real Safe

Reviewer: Btw do we use safe in testnet?

Author: Yes! We do this to be able to use the same recipes and test the runbooks on testnet.
If you drop me an address, I’ll add you.

Reviewer: `0x7B5Cc5B706FC4Cbe2ddd2C5D18Ece12A3E45878a`
